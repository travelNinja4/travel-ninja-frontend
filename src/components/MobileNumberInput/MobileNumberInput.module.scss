/**
------------------------------------------------------
styles specific to the MobileNumberInput component
------------------------------------------------------
**/

.container {
  display: flex;
  flex-direction: row;
  border: 1px solid var(--color-gray-300);
  border-radius: 8px;
  margin-top: 8px;
}

.label {
  font-weight: 600;
  font-size: 0.95rem;
  color: var(--color-gray-800);
  margin-bottom: 0.5rem;
}

.dropdownWrapper {
  position: relative;
  width: 6rem;
  border-right: 1px solid var(--color-gray-300);
}

.dropdown {
  border-right: 1px solid var(--color-gray-300);
  width: 6rem;
  border-radius: 0;
}

.dropdownText {
  border-radius: 8px;
  border: 0;
}

.input {
  flex: 1;
  border: none;
  border-radius: 8px;
}

.input:focus {
  outline: none;
}

.errorText {
  color: var(--color-red-500);
  font-size: 0.875rem;
  margin-top: 4px;
}
