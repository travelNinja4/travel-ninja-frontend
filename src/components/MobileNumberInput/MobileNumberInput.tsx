/**
 * A reusable input component for entering mobile numbers with a selectable country code and dynamic validation.
 *
 * @example
 * ```tsx
 * import MobileNumberInput from '@src/components/MobileNumberInput'
 *
 * export default function MobileNumberInput() {
 *   return <MobileNumberInput label="Hello" />;
 * }
 * ```
 */
'use client';

import { useEffect, useMemo, useState } from 'react';
import TextField from '../TextField';
import Autocomplete from '../Autocomplete';
import { useNotification } from '@/providers/NotificationProvider';
import { errorHandler } from '@/utils/errorHandler';
import clsx from 'clsx';
import styles from './MobileNumberInput.module.scss';
import { authService } from '@/services/authService';
import { useCommonStore } from '@/store/common';

/**
 * Define the props available for the MobileNumberInput component.
 */
interface MobileNumberInputProps {
  value?: { country: string; number: string };
  onChange?: (val: { country: string; number: string }) => void;
  maxLength?: number;
  error?: string;
  countryError?: string;
}

export default function MobileNumberInput({
  maxLength,
  error,
  countryError,
  value,
  onChange,
}: MobileNumberInputProps) {
  const { showNotification } = useNotification();
  const { countries, setCountries } = useCommonStore();
  const [country, setCountry] = useState('');
  const [search, setSearch] = useState('');
  const [number, setNumber] = useState(value?.number ?? '');
  const [length, setLength] = useState(
    value?.country ? countries.find((c) => c.code === value.country)?.length : maxLength,
  );

  useEffect(() => {
    (async () => {
      try {
        const response = await authService.getCountryCode();
        setCountries(response);
      } catch (err) {
        const messages = errorHandler(err);
        messages.forEach((msg) => showNotification('Error!', msg, 'error'));
      }
    })();
  }, []);

  useEffect(() => {
    if (countries.length === 0) return;

    // Set initial country and number if value exists
    if (value) {
      setCountry(value.country);
      setNumber(value.number);
      setLength(
        value.country ? countries.find((c) => c.code === value.country)?.length : maxLength,
      );
    }
  }, [countries, value, maxLength]);

  const filteredCountries = useMemo(() => {
    if (!search) return countries;
    return countries.filter(
      (c) =>
        c.code.toLowerCase().includes(search.toLowerCase()) ||
        c.name.toLowerCase().includes(search.toLowerCase()),
    );
  }, [countries, search]);

  const handleSelect = (
    selected: { name: string; code: string; length: number } | string | undefined,
  ) => {
    if (!selected || typeof selected === 'string') {
      // if undefined or a string, reset country
      setCountry('');
      setSearch('');
      setNumber('');
      setLength(maxLength);
      onChange?.({ country: '', number: '' });
      return;
    }

    // selected is the proper country object
    setCountry(selected.code);
    setSearch('');
    setNumber('');
    setLength(selected.length);
    onChange?.({ country: selected.code, number: '' });
  };

  const handleSearch = (val: string) => {
    setSearch(val);
    if (!val) {
      setCountry('');
    }
  };

  return (
    <>
      <div data-testid="MobileNumberInputTest" className={styles.container}>
        <div className={styles.dropdownWrapper}>
          <Autocomplete
            options={filteredCountries?.map((c) => ({ value: c, label: c.code }))}
            value={country}
            searchValue={search}
            onSearchChange={(val) => handleSearch(val)}
            onChange={handleSelect}
            dropdownContainerClassName={clsx(styles.dropdown)}
            dropdownTextContainerClassName={styles.dropdownText}
          />
        </div>

        <TextField
          placeholder="Enter mobile number"
          className={clsx(styles.input)}
          value={number}
          maxLength={length || maxLength}
          onChange={(e) => {
            const newNumber = e.target.value;
            setNumber(newNumber);
            onChange?.({ country, number: newNumber });
          }}
        />
      </div>
      {error && <p className={styles.errorText}>{error}</p>}
      {countryError && <p className={styles.errorText}>{countryError}</p>}
    </>
  );
}
