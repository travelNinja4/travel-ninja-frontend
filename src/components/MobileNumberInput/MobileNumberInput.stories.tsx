import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import MobileNumberInput from './MobileNumberInput'

const meta = {
  title: 'components/MobileNumberInput',
  component: MobileNumberInput,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: { 
        control: 'text', 
        description: 'Sample label for the component' 
        },
  },
  args: { 
    /** onClick: fn() **/
    },
} satisfies Meta<typeof MobileNumberInput>;

export default meta;
type Story = StoryObj<typeof meta>;