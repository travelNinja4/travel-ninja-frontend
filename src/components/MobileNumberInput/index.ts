import MobileNumberInput from './MobileNumberInput'

export default MobileNumberInput