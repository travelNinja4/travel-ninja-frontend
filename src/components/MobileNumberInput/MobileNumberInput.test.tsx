import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import MobileNumberInput from './MobileNumberInput';

describe('Home', () => {
  /** Base props for MobileNumberInput **/
  const baseProps = {
    value: { country: 'US', number: '1234567890' },
    onChange: jest.fn(),
    maxLength: 10,
    error: '',
    countryError: '',
  };

  /**
   * @test
   * Renders the MobileNumberInput component without throwing an error
   * and checks if the MobileNumberInputTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<MobileNumberInput {...baseProps} />);
    expect(screen.getByTestId('MobileNumberInputTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the MobileNumberInput component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<MobileNumberInput {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the MobileNumberInput component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<MobileNumberInput {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
