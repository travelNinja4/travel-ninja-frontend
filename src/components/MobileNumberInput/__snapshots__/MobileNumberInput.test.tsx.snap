// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="MobileNumberInputTest"
  >
    <div
      class="dropdownWrapper"
    >
      <div
        class="dropdown dropdown"
        data-testid="AutocompleteTest"
      >
        <div
          class="inputWrapper"
        >
          <input
            class="input dropdownText"
            placeholder="Country code"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="container input"
      data-testid="TextFieldTest"
    >
      <input
        class="inputField"
        maxlength="10"
        placeholder="Enter mobile number"
        type="text"
        value="1234567890"
      />
    </div>
  </div>
</DocumentFragment>
`;
